arch/arm/src/samv7/sam_pwm.c

```diff
@@ -493,71 +449,39 @@ static int pwm_start(struct pwm_lowerhalf_s *dev,
                      const struct pwm_info_s *info)
 {
   struct sam_pwm_s *priv = (struct sam_pwm_s *)dev;
-  int ret = OK;
 
-  /* Change frequency only if it is needed */
-
-  if (info->frequency != priv->frequency)
-    {
+#ifdef CONFIG_PWM_MULTICHAN
       for (int i = 0; i < PWM_NCHANNELS; i++)
         {
-#ifdef CONFIG_PWM_MULTICHAN
+          int8_t index = info->channels[i].channel;
+
           /* Break the loop if all following channels are not configured */
 
-          if (info->channels[i].channel == -1)
+          if (index == -1)
             {
               break;
             }
 
           /* Configure the module freq only if is set to be used */
 
-          if (info->channels[i].channel != 0)
+          if (index != 0)
             {
-              ret = pwm_change_freq(dev, info, i);
-            }
-#else
-          ret = pwm_change_freq(dev, info, i);
-#endif
-        }
-
-      /* Save current frequency */
-
-      if (ret == OK)
-        {
-          priv->frequency = info->frequency;
-        }
-    }
+              /* Set the frequency and enable PWM output for each channel */
 
-#ifdef CONFIG_PWM_MULTICHAN
-  for (int i = 0; ret == OK && i < PWM_NCHANNELS; i++)
-    {
-      /* Break the loop if all following channels are not configured */
-
-      if (info->channels[i].channel == -1)
-        {
-          break;
-        }
-
-      /* Enable PWM output for each channel */
-
-      if (info->channels[i].channel != 0)
-        {
-          ret = pwm_set_output(dev, info->channels[i].channel - 1,
-                                    info->channels[i].duty);
+              pwm_set_freq(dev, priv->channels[index - 1].channel,
```

Comment:
I am not sure about this. Lets say we configure only channels 2 and 3.

`int8_t index = info->channels[i].channel;` on line 456 gives as `index = 3` for channel 3. But then we would be out of `priv->channels[]` range with `priv->channels[3-1]` since only two channels are configured. There is no assurance all 4 channels will be configured.

I think the previous `info->channels[i].channel - 1` is sufficient here, all we need to do is just decrement the channel number used from application (from 1 to 4) so it would match register's offset (0 to 3).

The same goes for` pwm_set_output(dev, priv->channels[index - 1].channel, info->channels[i].duty);`.

I need to go now, I will be able to try the code in few hours.

Reply:
Yes. I missed the index check. Will add it.

Reply:
Range check added. Please try this patch

Reply:
The `info->channels` hold a logical PWM channels and those could be mapped to different HW PWM channels. I mean we can have `CONFIG_PWM_NCHANNELS` set to 2 and having `CONFIG_SAMV7_PWM0_CH2` and `CONFIG_SAMV7_PWM0_CH3` enabled. So `info->channels` will result to `info->channels[0].channel == 1` and `info->channels[1].channel == 2` and should be mapped to HW channels 2 and 3. Do you agree?

Reply:
I'm planning to add support of complementary output as well in the next couple of days

Reply:
Should not the mapping be vice versa?. Or the question is what kind of numbers are we putting into `info->channels`. The number of channel (from 1 to 4 - CH0 = 1, CH1=2 etc...) or the position number (first configured channel equals 1, second equals 2 etc.)?

My understanding is that if we have config options `CONFIG_SAMV7_PWM0_CH0` to `CONFIG_SAMV7_PWM0_CH3` and we select `CONFIG_SAMV7_PWM0_CH2` and `CONFIG_SAMV7_PWM0_CH3` then channels' numbers put into `info->channels` should be 3 and 4 (we cannot use 0 so we need to index from 1 and this is third and fourth channel in SAM's PWM). It is better than mapping first and second channel from my point of view because user does not need to check what other channels are configured. I wasnt able to find any standard aproach in the documentation thought.

Reply:
Yeah, I wasn't able to find any description as well. but my points are next:
1. The application (for example `examples/pwm`)  should be portable from board to board and across the chips (again this is purely my understanding) so it has to operate with "virtual" channels like it does now:
```
config EXAMPLES_PWM_CHANNEL1
	int "First PWM channel number"
	default 1
	range 1 6
	---help---
		The first PWM channel number.  Default: 1
```
2. Each specific board can implement PWM "virtual" channel based on specific HW details (map to HW channel 3 or 4 or even 567).

It is like with `/dev/timerX`, so `/dev/time1` does not mean that HW timer 1 is used.

But anyway let's move this discussion to the mailing thread.

Reply:
I agree with your point about board to board (chip to chip) portability which is better with virtual channels. I tested your branch and managed to get both single channel and multiple channels output (using virtual channels in `example/pwm` app).

Yes, we can move the discussion regarding channel indexes to mailing list. I am now inclining towards virtual channels but we should document it properly somewhere.

Reply:
what's the final decision?

Reply:
I wrote this to mailing list, but no response till now. I think we need to make a clear description of how PWM multichannel works and update all the drivers accordingly.

Reply:
I think we can merge this if there is no interest in mailing list. It would be great to standardized the PWM usage some day but I think it should not dealy this fix any further.